### PR TITLE
chore: cheat code documentation

### LINKED ISSUE #300: [testing] add details on cairo test runner cheatcodes
Currently Dojo is using Cairo test runner, where [few cheat codes](https://github.com/starkware-libs/cairo/blob/25733b7a57b5f43b08b7d61f947b0396b3743801/corelib/src/starknet/testing.cairo#L34) are available, but not much documented.

We should add in the testing chapter a section/page to explain each of them with examples.

### PR DESCRIPTION
**Pull request by od-hunter**
resolves #300

### CODE REVIEW

File: docs/pages/framework/testing-cheat-codes.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+assert(get_block_timestamp == 123456, 'bad timestamp');
+
+```
+
+5. `set_version`
+
+This cheat code helps one set the version to the provided value, enabling you to test contract behavior with different versions. You can apply this when:
+
+- Testing a contract's behavior with different versions, such as checking if a certain function is only callable in a specific version.
+- Simulating a scenario where a contract is upgraded to a different version.
```
Comment (glihm) [inline docs/pages/framework/testing-cheat-codes.md:72]: this cheat code is about the version of the transaction, not the contract's version.
Could you please rephrase accordingly?

File: docs/pages/framework/testing-cheat-codes.md [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+```
+use starknet::{testing, get_version};
```
Comment (glihm) [inline docs/pages/framework/testing-cheat-codes.md:75]:
```suggestion
use starknet::{testing, get_tx_info};
```
`get_version` function doesn't exist. It's the version of the transaction inside the transaction execution info:
```rust
testing::set_version(1);
assert_eq!(get_tx_info().unbox().version, 1_felt252);
```

File: docs/pages/framework/testing-cheat-codes.md [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+assert(get_version == '0.1.0', 'wrong version')
+
+```
+
+6. `set_account_contract_address`
+
+This cheat code helps you set the account contract address to the provided value, allowing you to test contract interactions with different account contracts. You can apply this when:
+
+- Testing a contract's interaction with a specific account contract, such as a wallet contract.
```
Comment (glihm) [inline docs/pages/framework/testing-cheat-codes.md:87]:
```suggestion
- Testing a contract's interaction with a specific account contract.
```

File: docs/pages/framework/testing-cheat-codes.md [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+- Simulating a scenario where a contract is called by a different account contract.
+
+```
+use starknet::{testing, get_account_contract_address, contract_address_const};
+
+const contract: ContractAddress = contract_address_const::<'contract'>();
```
Comment (glihm) [inline docs/pages/framework/testing-cheat-codes.md:93]:
```suggestion
const contract = contract_address_const::<'contract'>();
```
Type is inferred, it makes the code more readable. 👍

File: docs/pages/framework/testing-cheat-codes.md [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+6. `set_account_contract_address`
+
+This cheat code helps you set the account contract address to the provided value, allowing you to test contract interactions with different account contracts. You can apply this when:
+
+- Testing a contract's interaction with a specific account contract, such as a wallet contract.
+- Simulating a scenario where a contract is called by a different account contract.
+
+```
+use starknet::{testing, get_account_contract_address, contract_address_const};
```
Comment (glihm) [inline docs/pages/framework/testing-cheat-codes.md:91]:
```suggestion
use starknet::{testing, get_tx_info, contract_address_const};
```
Same here, this function doesn't exist.

File: docs/pages/framework/testing-cheat-codes.md [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+
+const contract: ContractAddress = contract_address_const::<'contract'>();
+
+testing::set_account_contract_address(contract);
+
+assert(get_account_contract_address() == contract, 'wrong contract address');
```
Comment (glihm) [inline docs/pages/framework/testing-cheat-codes.md:97]:
```suggestion
assert_eq!(get_tx_info().unbox().account_contract_address.into(), contract);
```
Using `assert_eq!` you don't always require to put a reason. Can be sometimes shorter and better. :)

**Review by glihm — CHANGES_REQUESTED** [review]
Thanks for the revision @od-hunter. Most of the examples are still not using real functions from the core lib. I've made some corrections but not all, could you please take the time to check if all the examples compile and make the according corrections with the examples given?

Like so, the reader of the book will have accurate information. 🫡

File: docs/pages/framework/testing-cheat-codes.md [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+
+```
+use starknet::{testing};
+let contract_address = starknet::contract_address_const::<0x42>();
+
+testing::pop_l2_to_l1_message(contract_address)
+
+```
+
+In conclusion, the Cairo cheat codes provide a powerful toolset for testing and debugging Starknet contracts. By mastering these cheat codes, you can simulate various scenarios, test edge cases, and ensure the correctness of your contracts. Remember to use them wisely and in conjunction with other testing techniques to achieve comprehensive coverage. With this guide, you are now well-equipped to tackle complex testing challenges and build robust contracts on the Starknet network. Happy testing!
```
Comment (glihm) [inline docs/pages/framework/testing-cheat-codes.md:229]:
```suggestion
In conclusion, the Cairo cheat codes provide a powerful toolset for testing and debugging Starknet contracts. By mastering these cheat codes, you can simulate various scenarios, test edge cases, and ensure the correctness of your contracts. With this guide, you are now well-equipped to tackle dojo testing which heavily uses those cheat codes.
```

**Review by glihm — CHANGES_REQUESTED** [review]
@od-hunter thanks for the work here, in a good direction but not yet ready to be published.

I've prepared for you a test module that you can run to test all the cheat codes and adjust the documentation accordingly. Some of the examples are not working, could you please adjust with the given test module to ensure everything compiles? Thank you very much!

```rust
#[cfg(test)]
mod tests {
    use starknet::{testing, get_block_number, get_caller_address, get_contract_address, contract_address_const, get_block_timestamp, get_tx_info};

    #[test]
    fn f1() {
        testing::set_block_number(1234567);
        assert!(get_block_number() == 1234567, "bad block number");
    }

    #[test]
    fn f2() {
        let user_one = contract_address_const::<'user1'>();
        testing::set_caller_address(user_one);
        assert!(get_caller_address() == user_one, "bad caller");
    }

    #[test]
    fn f3() {
        let hub_address = contract_address_const::<'hub'>();
        testing::set_contract_address(hub_address);
        assert!(get_contract_address() == hub_address, "bad address");
    }

    #[test]
    fn f4() {
        testing::set_block_timestamp(123456);
        assert!(get_block_timestamp() == 123456, "bad timestamp");
    }

    #[test]
    fn f5() {
        testing::set_version('0.1.0');
        assert_eq!(get_tx_info().unbox().version, '0.1.0');
    }

    #[test]
    fn f6() {
        let contract = contract_address_const::<'contract'>(); 
        testing::set_account_contract_address(contract);
        assert_eq!(get_tx_info().unbox().account_contract_address.into(), contract);
    }

    #[test]
    fn f7() {
        testing::set_max_fee(123456);
        assert_eq!(get_tx_info().unbox().max_fee.into(), 123456);
    }

    #[test]
    fn f8() {
        testing::set_transaction_hash('12345678');
        assert_eq!(get_tx_info().unbox().transaction_hash, '12345678');
    }

    #[test]
    fn f9() {
        testing::set_chain_id('test_chain_id');
        assert_eq!(get_tx_info().unbox().chain_id.into(), 'test_chain_id');
    }

    #[test]
    fn f10() {
        testing::set_nonce('test_nonce');
        assert_eq!(get_tx_info().unbox().nonce, 'test_nonce');
    }

    #[test]
    fn f11() {
        testing::set_signature(array!['r', 's'].span());
        let s = get_tx_info().unbox().signature;
        assert_eq!(*s[0], 'r');
        assert_eq!(*s[1], 's');
    }

    #[test]
    fn f12() {
        testing::set_block_hash(12345678, 'value');
        assert_eq!(get_tx_info().unbox().block_hash, 12345678);
    }

    #[test]
    fn f13() {
        let contract_address = starknet::contract_address_const::<0x42>();
        let _log = testing::pop_log_raw(contract_address);
    }

    #[derive(Drop, starknet::Event)]
    struct TestEvent {
        #[key]
        id: u32,
        data: felt252,
    }

    #[test]
    fn f14() {
        let contract_address = starknet::contract_address_const::<0x42>();
        let _log = testing::pop_log::<TestEvent>(contract_address);
    }

    #[test]
    fn f15() {
        let contract_address = starknet::contract_address_const::<0x42>();
        let _msg = testing::pop_l2_to_l1_message(contract_address);
    }
}
```

**Comment by od-hunter** [comment]
gm sir
i am a bit stuck in the last 3 there test examples: `pop_log_raw`, `pop_log`, `pop_l2_to_l1_message`

i don’t know how to go about writing the assert statments for them 
please can you assist me with resources to solve it

**Comment by glihm** [comment]
> gm sir @glihm i am a bit stuck in the last 3 there test examples: `pop_log_raw`, `pop_log`, `pop_l2_to_l1_message`
> 
> i don’t know how to go about writing the assert statments for them please can you assist me with resources to solve it

Sorry for the delay here.

For the messages, you can skip them for now. These are not used that much for the moment in Dojo.

In the meantime, the `pop_log` is very used and important.
To test it correctly, you should have a contract that actually emits an event. And then the event will be caught by this cheat code.

What you can do to make the full test, write a minimal contract that emits an event, call the function that emits the event, and then pop the log.

A minimal contract could look something like this:
```rust
#[starknet::interface]
pub trait IContract<T> {
    fn emit_log(ref self: T);
}

#[starknet::contract]
pub mod MyContract {
    #[abi(embed_v0)]
    impl IContractImpl of IContract<ContractState> {
        fn emit_log(ref self: ContractState) {
            starknet::emit(.....);
        }
    }
}
```
You can use the demo event, don't forget to declare inside the contract. 👍

**Comment by od-hunter** [comment]
@glihm , please review

**Review by glihm — APPROVED** [review]
Thank you @od-hunter for the work done here!

**Comment by od-hunter** [comment]
Thank you sir @glihm , I’m open to more issues🚀🚀
